services/device-registry-file/src/main/java/org/eclipse/hono/deviceregistry/file/FileBasedRegistrationService.java

```diff
@@ -520,6 +526,102 @@ private JsonObject convertDevice(final String deviceId, final Device payload) {
+     * so the paging may be inconsistent.
+     */
+    @Override
+    public Future<OperationResult<List<DeviceWithId>>> searchDevices(final String tenantId, final Optional<Integer> limit, final Optional<Integer> offset,
+                                                                     final Optional<List<SortingOption>> sorting, final Optional<List<Filter>> filters, final Span span) {
+
+        final int actualLimit = limit.orElse(config.getDefaultPageLimit());
+        final int actualOffset = offset.orElse(0);
+
+        return Future.succeededFuture(doSearchDevices(tenantId, actualLimit, actualOffset, sorting.orElse(null), filters.orElse(null)));
+    }
+
+    private OperationResult<List<DeviceWithId>> doSearchDevices(final String tenantId, final int limit, final int offset,
+                                                                final List<SortingOption> sort, final List<Filter> filters) {
+
+        final Set allDevicesIds = getDevicesForTenant(tenantId).keySet();
+        List<DeviceWithId> devicesList = new ArrayList<>();
+
+        for (int i = offset; i < limit + offset; i++ ) {
+            final String id = (String) allDevicesIds.toArray()[i];
+            final Device device = getDevicesForTenant(tenantId).get(id).getValue();
+            final DeviceWithId withId = DeviceWithId.fromDevice(device, id);
+
+            devicesList.add(withId);
+        }
+
+        for (Filter f : filters) {
+            devicesList = filterDeviceList(devicesList, f);
+        }
+
+        for (SortingOption s : sort) {
+            devicesList = sortDeviceList(devicesList, s);
+        }
+
+        return OperationResult.ok(HttpURLConnection.HTTP_OK,
+                devicesList,
+                Optional.ofNullable(DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())),
+                Optional.empty());
+    }
+
+    private List<DeviceWithId> filterDeviceList(final List<DeviceWithId> list, final Filter filter) {
+
+        return list
+                .stream()
+                .filter(buildJsonBasedPredicate(filter))
+                .collect(Collectors.toList());
+    }
+
+    //TODO : support JSON Path
+    // won't be supported by vertx : https://github.com/eclipse-vertx/vert.x/issues/1661
+    private Predicate<DeviceWithId> buildJsonBasedPredicate(final Filter filter) {
+        return deviceWithId -> {
+            final Map<String, Object> device = JsonObject.mapFrom(deviceWithId).getMap();
+
+            switch (filter.getOperation()) {
+                case eq:
+                    return device.get(filter.getField()) == filter.getValue();
+                default:
+                    return true;
+            }
+        };
+    }
+
+    private List<DeviceWithId> sortDeviceList(final List<DeviceWithId> list, final SortingOption sort) {
+
+        return list
+                .stream()
+                .sorted(buildJsonBasedComparator(sort))
```

Comment:
Same here, if we could build one comparator and then use one stream() to filter and sort that would be great.

Reply:
I am not sure how that one is possible.